### DISCUSSION TITLE
next/image and imgix using signed urls

### DISCUSSION THREAD

**Discussion by tbrolin** [vercel/next.js#43521] [Help] [answered]
Is it even possible?

The urls must be "signed" using a secret key and it has to be done after all parameters are known, this means the built in loader for imgix is a no go. I tried using a custom loader, however it seems like the loader function generating urls is run both server and client side, but using the secret key on the client is a no go.

Clients upload and publish images through a cms, so there is not really any way to know what images is needed in beforehand.

I couldn't find any good information about this problem on the internet either so if someone is a pro-googler and find something, those answers are also appreciated.

Someone have any experience of using imgix signed urls with next/image?

**Comment by wwwdepot** [comment]
You need to do these in your CMS

1. Client upload image in your CMS
2. A "create or update" event webhook goes off and reach a specific endpoint
3. The endpoint will sign the url & update CMS image object with signed URL
4. Nextjs will use signed URL directly

OR sign the URLs on-the-fly (not recommended, but you can):

1. fetch data in Next.js
2. look for URLs need to be signed & sign those URLs
3. return modified props to next.js
4. Nextjs will use signed URL directly

**Update**: maybe you can use https://nextjs.org/docs/api-reference/next/image#loader-configuration

**Comment by tbrolin** [comment, marked as answer]
Thanks for the reply!

I should point out that we are using a cms with a headless approach.

I am afraid that the first option is sort of the culprit, it would be nice to just get a url to an image, however, next/image generates new urls based on whatever "sizes" you configured, this will result in five new urls each of which needs to be signed individually.

The last option using the shipped imgix loader failed because it does not support signing of urls. And that is probably because of the same reason I couldn't create a custom loader for imgix, the loader runs both on the server and the client, and signing urls requires a secret key that you must not expose client wise.

So I was left with a variant of the second option you propose. I still would need a custom loader to feed the next/image component with the signed urls. This leads to either having to know exactly where images are in the data fetched or inventing some nifty dsl to be able to detect what properties contains image urls. Both solutions are pretty ugly.

Luckily I managed to convince the allmighties to not use signed imgix urls at all, until like Next 14(?) where you hopefully actually can solve this using server components (who knows?)...

This made all complexity and anxiety I had in this matter vanish into thin air! :-D

Thanks anyways!
